internal/soc/soc.go

```diff
@@ -0,0 +1,80 @@
+// This source file is part of the EdgeDB open source project.
+//
+// Copyright 2020-present EdgeDB Inc. and the EdgeDB authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package soc has utilities for working with sockets.
+package soc
+
+import (
+	"net"
+	"reflect"
+	"unsafe"
+)
+
+const minChunkSize = 5
+
+// Data is the bytes & error that were read from a socket.
+// When data's bytes are no longer used data must be released.
+type Data struct {
+	Buf     []byte
+	Err     error
+	release func()
+}
+
+// Release frees the underlying array to be reused.
+func (d *Data) Release() {
+	if d.release != nil {
+		d.release()
+	}
+}
+
+// Read reads a socket sending the read data to toBeDeserialized.
+// This should be run in it's own go routine.
+func Read(conn net.Conn, freeMemory *MemPool, toBeDeserialized chan *Data) {
+	mkRelease := func(slab []byte) func() {
+		return func() {
+			freeMemory.Release(slab)
+		}
+	}
+
+	for {
+		slab := freeMemory.Acquire()
+		buf := slab
+
+		for len(buf) >= minChunkSize {
+			n, err := conn.Read(buf)
```

Comment:
Why don't you check `err` right here?

It looks like this goroutine is always reading data, right? So at some time the socket connection can be closed, in which case we just need to exit immediately, basically with check like this:


```
if err != nil {
  if conn.closed {
    return
  } else {
    toBeDeserialized <- &Data{Err: err}
    return
  }
}
```

Reply:
if there is an error and the connection is closed, the upstream go routine needs to get the error too or it will block forever waiting for data.

This sends data before checking the error because [the go docs say](https://golang.org/pkg/io/#Reader):
> Callers should always process the n > 0 bytes returned before considering the error err.